ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Block/SupportsDiffusionPipelining.hs

```diff
@@ -0,0 +1,230 @@
+  , NoThunks (TentativeHeaderState blk)
+  ) => BlockSupportsDiffusionPipelining blk where
+  -- | State that is maintained to judge whether a header can be pipelined. It
+  -- can be thought of as a summary of all past trap tentative headers.
+  type TentativeHeaderState blk :: Type
+
+  -- | The initial 'TentativeHeaderState'. This is used as the initial value on
+  -- node startup, as well as by the HFC instance for new eras.
+  initialTentativeHeaderState :: Proxy blk -> TentativeHeaderState blk
+
+  -- | Apply a header to the 'TentativeHeaderState'. This returns @'Just' st@ to
+  -- indicate that the header can be pipelined, and that the
+  -- 'TentativeHeaderState' must be updated to @st@ if the header turns/turned
+  -- out to be a trap header (the corresponding block body is invalid).
+  updateTentativeHeaderState ::
+       BlockConfig blk
+    -> Header blk
+       -- ^ A (valid) header whose block body is either not yet known to be
+       -- valid, or definitely invalid.
+    -> TentativeHeaderState blk
+       -- ^ The most recent 'TentativeHeaderState' in this particular context.
+    -> Maybe (TentativeHeaderState blk)
+       -- ^ The new 'TentativeHeaderState' in case the header satisfies the
+       -- pipelining criterion and is a trap header.
+
+{-------------------------------------------------------------------------------
+  DerivingVia helpers
+-------------------------------------------------------------------------------}
+
+-- | A @DerivingVia@ helper to implement 'BlockSupportsDiffusionPipelining' for
+-- blocks where no header should ever be pipelined.
+--
+-- > deriving via DisableDiffusionPipelining MyBlock
+-- >   instance BlockSupportsDiffusionPipelining MyBlock
+newtype DisableDiffusionPipelining blk = DisableDiffusionPipelining blk
+
+newtype instance Header (DisableDiffusionPipelining blk) =
+  DisableDiffusionPipeliningHeader (Header blk)
+
+newtype instance BlockConfig (DisableDiffusionPipelining blk) =
+  DisableDiffusionPipeliningBlockConfig (BlockConfig blk)
+
+instance BlockSupportsDiffusionPipelining (DisableDiffusionPipelining blk) where
+  type TentativeHeaderState _ = ()
+
+  initialTentativeHeaderState _ = ()
+
+  updateTentativeHeaderState _cfg _hdr () =
+      Nothing
+
+-- | A @DerivingVia@ helper to implement 'BlockSupportsDiffusionPipelining' for
+-- blocks where a header should be pipelined iff it has a better 'SelectView'
+-- than the last tentative trap header.
+--
+-- > deriving via DisableDiffusionPipelining MyBlock
+-- >   instance BlockSupportsProtocol blk
+-- >   => BlockSupportsDiffusionPipelining MyBlock
+--
+-- This requires that the 'SelectView' is totally ordered, in particular that
+-- the order is transitive.
+--
+-- For example, if @'SelectView' ~ 'BlockNo'@, this means that a header can be
+-- pipelined if it has a larger block number than the last tentative trap
+-- header. So if someone diffused a trap header for a particular block height,
```

Comment:
"So if ..." it'd be helpful to elaborate to something like "This would violate "Usefulness despite adversarial activity", because ..."